File: leafy/Services/Cloudflare/MyLeafyBackendClient.swift
import Foundation
import Security

nonisolated struct MyLeafyBackendSession: Codable, Sendable {
    let token: String
    let userID: UUID
    let isAnonymous: Bool
}

nonisolated struct MyLeafyBackendError: LocalizedError, Sendable {
    let status: Int
    let code: String
    let message: String
    let requestID: String?

    var errorDescription: String? { message }
}

nonisolated protocol MyLeafyBackendSessionStoring: Sendable {
    func load() throws -> MyLeafyBackendSession?
    func save(_ session: MyLeafyBackendSession) throws
    func remove() throws
}

nonisolated struct MyLeafyBackendKeychainStore: MyLeafyBackendSessionStoring {
    let origin: String

    private var query: [String: Any] {
        [kSecClass as String: kSecClassGenericPassword,
         kSecAttrService as String: "com.myleafy.cloudflare-session",
         kSecAttrAccount as String: origin,
         kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly]
    }

    func load() throws -> MyLeafyBackendSession? {
        var lookup = query
        lookup[kSecReturnData as String] = true
        lookup[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: CFTypeRef?
        let status = SecItemCopyMatching(lookup as CFDictionary, &result)
        if status == errSecItemNotFound { return nil }
        guard status == errSecSuccess, let data = result as? Data else { throw storageError(status) }
        return try JSONDecoder().decode(MyLeafyBackendSession.self, from: data)
    }

    func save(_ session: MyLeafyBackendSession) throws {
        let data = try JSONEncoder().encode(session)
        let status = SecItemUpdate(query as CFDictionary, [kSecValueData as String: data] as CFDictionary)
        if status == errSecSuccess { return }
        guard status == errSecItemNotFound else { throw storageError(status) }
        var insertion = query
        insertion[kSecValueData as String] = data
        let inserted = SecItemAdd(insertion as CFDictionary, nil)
        guard inserted == errSecSuccess else { throw storageError(inserted) }
    }

    func remove() throws {
        let status = SecItemDelete(query as CFDictionary)
        guard status == errSecSuccess || status == errSecItemNotFound else { throw storageError(status) }
    }

    private func storageError(_ status: OSStatus) -> MyLeafyBackendError {
        MyLeafyBackendError(status: 0, code: "session_storage_\(status)", message: "无法安全保存登录状态，请稍后重试。", requestID: nil)
    }
}

/// URLSession transport for the versioned Workers API. Construction performs no
/// network request; guest mode can keep using local stores without creating it.
actor MyLeafyBackendClient {
    nonisolated let baseURL: URL
    private let network: URLSession
    private let storage: any MyLeafyBackendSessionStoring
    private var establishingSession: Task<Void, Error>?
    private var currentSession: MyLeafyBackendSession?

    init(baseURL: URL, network: URLSession? = nil, storage: (any MyLeafyBackendSessionStoring)? = nil) throws {
        guard baseURL.scheme == "https", baseURL.host != nil, baseURL.user == nil, baseURL.password == nil,
              baseURL.query == nil, baseURL.fragment == nil, baseURL.path.isEmpty || baseURL.path == "/" else {
            throw MyLeafyBackendError(status: 0, code: "invalid_api_origin", message: "后台地址配置无效。", requestID: nil)
        }
        self.baseURL = baseURL
        if let network {
            self.network = network
        } else {
            let configuration = URLSessionConfiguration.ephemeral
            configuration.httpCookieStorage = nil
            configuration.httpShouldSetCookies = false
            configuration.urlCache = nil
            self.network = URLSession(configuration: configuration)
        }
        let resolvedStorage = storage ?? MyLeafyBackendKeychainStore(origin: baseURL.absoluteString)
        self.storage = resolvedStorage
        self.currentSession = try resolvedStorage.load()
    }

    func requireBearerToken() throws -> String {
        guard let currentSession else { throw MyLeafyBackendError(status: 401, code: "unauthenticated", message: "请重新登录。", requestID: nil) }
        return currentSession.token
    }

    var authUserID: UUID? { currentSession?.userID }

    func establishSession(legacyToken: String? = nil, legacyAnonymous: Bool = true) async throws {
        if currentSession != nil { return }
        if let establishingSession { return try await establishingSession.value }
        let task = Task {
            if let legacyToken { _ = try await exchangeLegacySession(accessToken: legacyToken, isAnonymous: legacyAnonymous) }
            else { try await anonymousSession() }
        }
        establishingSession = task
        defer { establishingSession = nil }
        try await task.value
    }

    func anonymousSession() async throws {
        if currentSession != nil { return }
        let (_, response, data) = try await send(path: "/v1/auth/sign-in/anonymous", method: "POST", body: Data("{}".utf8), authenticated: false)
        try captureSession(response: response, data: data)
    }

    func signIn(email: String, password: String) async throws -> UUID {
        let data = try JSONEncoder().encode(EmailPassword(email: email, password: password))
        let (_, response, payload) = try await send(path: "/v1/auth/sign-in/email", method: "POST", body: data, authenticated: false)
        try captureSession(response: response, data: payload)
        return currentSession!.userID
    }

    func signUp(email: String, password: String) async throws {
        struct Input: Encodable { let email: String; let password: String; let name: String }
        _ = try await send(path: "/v1/auth/sign-up/email", method: "POST", body: JSONEncoder().encode(Input(email: email, password: password, name: email)), authenticated: false)
    }

    func resendVerification(email: String) async throws {
        struct Input: Encodable { let email: String; let type = "email-verification" }
        _ = try await send(path: "/v1/auth/email-otp/send-verification-otp", method: "POST", body: JSONEncoder().encode(Input(email: email)), authenticated: false)
    }

    func verifyRegistration(email: String, otp: String) async throws -> UUID {
        struct Input: Encodable { let email: String; let otp: String }
        let (_, response, payload) = try await send(path: "/v1/auth/email-otp/verify-email", method: "POST", body: JSONEncoder().encode(Input(email: email, otp: otp)), authenticated: false)
        try captureSession(response: response, data: payload)
        return currentSession!.userID
    }

    func exchangeLegacySession(accessToken: String, isAnonymous: Bool) async throws -> UUID {
        struct Input: Encodable, Sendable { let access_token: String }
        struct Output: Decodable, Sendable { let token: String; let user_id: UUID }
        let payload = try JSONEncoder().encode(Input(access_token: accessToken))
        let (_, _, data) = try await send(path: "/v1/session/exchange", method: "POST", body: payload, authenticated: false)
        let output = try JSONDecoder().decode(Output.self, from: data)
        let session = MyLeafyBackendSession(token: output.token, userID: output.user_id, isAnonymous: isAnonymous)
        try storage.save(session)
        currentSession = session
        return session.userID
    }

    func signOut(localOnly: Bool = false) async throws {
        if !localOnly, currentSession != nil {
            _ = try await send(path: "/v1/auth/sign-out", method: "POST", body: Data("{}".utf8), authenticated: true)
        }
        try storage.remove()
        currentSession = nil
    }

    func get<Response: Decodable & Sendable>(_ path: String, query: [URLQueryItem] = [], authenticated: Bool = true) async throws -> Response {
        let (_, _, data) = try await send(path: path, query: query, method: "GET", authenticated: authenticated)
        return try Self.decoder().decode(Response.self, from: data)
    }

    func request<Response: Decodable & Sendable, Body: Encodable & Sendable>(
        _ path: String, method: String = "POST", body: Body
    ) async throws -> Response {
        let (_, _, data) = try await send(path: path, method: method, body: JSONEncoder().encode(body), authenticated: true)
        return try Self.decoder().decode(Response.self, from: data)
    }

    func perform<Body: Encodable & Sendable>(_ path: String, method: String = "POST", body: Body) async throws {
        _ = try await send(path: path, method: method, body: JSONEncoder().encode(body), authenticated: true)
    }

    func uploadRequest(kind: String, uploadID: UUID, postID: UUID?, name: String, contentType: String) throws -> URLRequest {
        guard let session = currentSession else { throw MyLeafyBackendError(status: 401, code: "unauthenticated", message: "请重新登录。", requestID: nil) }
        var url = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
        url.path = "/v1/files/upload"
        url.queryItems = [URLQueryItem(name: "kind", value: kind), URLQueryItem(name: "upload_id", value: uploadID.uuidString.lowercased()), URLQueryItem(name: "name", value: name)]
        if let postID { url.queryItems?.append(URLQueryItem(name: "post_id", value: postID.uuidString.lowercased())) }
        var request = URLRequest(url: url.url!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(session.token)", forHTTPHeaderField: "Authorization")
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        return request
    }

    nonisolated func changeEvents(scope: String) -> AsyncThrowingStream<Void, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let socket = try await self.events(scope: scope)
                    socket.resume()
                    defer { socket.cancel(with: .goingAway, reason: nil) }
                    try await withTaskCancellationHandler {
                        while !Task.isCancelled {
                            let message = try await socket.receive()
                            let data: Data
                            switch message {
                            case .data(let bytes): data = bytes
                            case .string(let string): data = Data(string.utf8)
                            @unknown default: continue
                            }
                            struct Event: Decodable { let type: String }
                            if try JSONDecoder().decode(Event.self, from: data).type == "changed" { continuation.yield(()) }
                        }
                    } onCancel: { socket.cancel(with: .goingAway, reason: nil) }
                    continuation.finish()
                } catch {
                    if Task.isCancelled { continuation.finish() } else { continuation.finish(throwing: error) }
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    func upload<Response: Decodable & Sendable>(data: Data, contentType: String, query: [URLQueryItem]) async throws -> Response {
        let (_, _, response) = try await send(path: "/v1/files/upload", query: query, method: "POST", body: data, contentType: contentType, authenticated: true)
        return try Self.decoder().decode(Response.self, from: response)
    }

    func download(bucket: String, path: String) async throws -> Data {
        let (_, _, data) = try await send(path: "/v1/files/read", query: [URLQueryItem(name: "bucket", value: bucket), URLQueryItem(name: "path", value: path)], method: "GET", authenticated: true)
        return data
    }

    func events(scope: String) throws -> URLSessionWebSocketTask {
        guard scope == "feed" || scope == "notifications", let session = currentSession else {
            throw MyLeafyBackendError(status: 401, code: "unauthenticated", message: "请重新登录。", requestID: nil)
        }
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
        components.scheme = "wss"
        components.path = "/v1/events/\(scope)"
        var request = URLRequest(url: components.url!)
        request.setValue("Bearer \(session.token)", forHTTPHeaderField: "Authorization")
        return network.webSocketTask(with: request)
    }

    private func send(path: String, query: [URLQueryItem] = [], method: String, body: Data? = nil,
                      contentType: String = "application/json", authenticated: Bool) async throws -> (Int, HTTPURLResponse, Data) {
        guard path.hasPrefix("/v1/") else {
            throw MyLeafyBackendError(status: 0, code: "invalid_api_path", message: "后台请求路径无效。", requestID: nil)
        }
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
        components.path = path
        components.queryItems = query.isEmpty ? nil : query
        var request = URLRequest(url: components.url!, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 30)
        request.httpMethod = method
        request.httpBody = body
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        if authenticated {
            guard let session = currentSession else {
                throw MyLeafyBackendError(status: 401, code: "unauthenticated", message: "请重新登录。", requestID: nil)
            }
            request.setValue("Bearer \(session.token)", forHTTPHeaderField: "Authorization")
        }
        let (data, response) = try await network.data(for: request)
        guard let response = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard (200..<300).contains(response.statusCode) else {
            if authenticated && response.statusCode == 401 && request.value(forHTTPHeaderField: "Authorization") == currentSession.map({ "Bearer \($0.token)" }) {
                try storage.remove()
                currentSession = nil
            }
            let failure = try? JSONDecoder().decode(Failure.self, from: data)
            throw MyLeafyBackendError(status: response.statusCode, code: failure?.errorEnvelope?.code ?? failure?.code ?? "http_\(response.statusCode)",
                                     message: failure?.errorEnvelope?.message ?? failure?.error ?? failure?.message ?? "后台请求失败，请稍后重试。",
                                     requestID: response.value(forHTTPHeaderField: "X-Request-ID"))
        }
        return (response.statusCode, response, data)
    }

    private func captureSession(response: HTTPURLResponse, data: Data) throws {
        try Task.checkCancellation()
        let result = try JSONDecoder().decode(AuthenticationResult.self, from: data)
        // Better Auth's bearer plugin returns a signed token in this header. Its
        // unsigned JSON token must never be used as the Authorization credential.
        guard let token = response.value(forHTTPHeaderField: "set-auth-token"), !token.isEmpty else {
            throw MyLeafyBackendError(status: 0, code: "missing_signed_session", message: "登录状态未建立，请重试。", requestID: nil)
        }
        let session = MyLeafyBackendSession(token: token, userID: result.user.id, isAnonymous: result.user.isAnonymous ?? false)
        try storage.save(session)
        currentSession = session
    }

    private static func decoder() -> JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { input in
            let value = try input.singleValueContainer().decode(String.self)
            let fractional = ISO8601DateFormatter()
            fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = fractional.date(from: value) ?? ISO8601DateFormatter().date(from: value) { return date }
            throw DecodingError.dataCorrupted(.init(codingPath: input.codingPath, debugDescription: "Invalid API timestamp"))
        }
        return decoder
    }

    private struct EmailPassword: Encodable { let email: String; let password: String }
    private struct AuthenticationResult: Decodable { let user: User }
    private struct User: Decodable { let id: UUID; let isAnonymous: Bool? }
    private struct Failure: Decodable {
        let error: String?; let code: String?; let message: String?; let errorEnvelope: Envelope?
        struct Envelope: Decodable { let code: String; let message: String }
    }
}

File: android/app/src/main/java/com/myleafy/android/services/cloudflare/MyLeafyBackendClient.kt
package com.myleafy.android.services.cloudflare

import com.myleafy.android.core.security.SecureStorage
import java.io.IOException
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

@Serializable
data class BackendSession(val token: String, val userId: String, val anonymous: Boolean)

interface BackendSessionStore {
    fun load(): BackendSession?
    fun save(session: BackendSession)
    fun clear()
}

class KeystoreBackendSessionStore(
    private val secure: SecureStorage,
    origin: String,
) : BackendSessionStore {
    private val key = "cloudflare-session:$origin"
    private val json = Json { ignoreUnknownKeys = true }

    override fun load(): BackendSession? {
        val raw = secure.read(key)
        if (raw == null && secure.has(key)) throw IOException("无法读取已保存的登录状态，请重新登录。")
        return raw?.let { json.decodeFromString<BackendSession>(it) }
    }

    override fun save(session: BackendSession) = secure.save(key, json.encodeToString(session))
    override fun clear() = secure.remove(key)
}

class BackendException(val status: Int, val code: String, override val message: String, val requestId: String?) : IOException(message)

/** Created only for an enabled online identity. No network I/O occurs in the constructor. */
class MyLeafyBackendClient(
    origin: String,
    private val store: BackendSessionStore,
    private val http: OkHttpClient = OkHttpClient.Builder()
        .callTimeout(30, TimeUnit.SECONDS)
        .followRedirects(false)
        .followSslRedirects(false)
        .build(),
) {
    private val originUrl = origin.toHttpUrl().also {
        require(it.isHttps && it.username.isEmpty() && it.password.isEmpty() && it.encodedPath == "/" && it.query == null && it.fragment == null) {
            "Backend origin must be an HTTPS origin without credentials, path or query."
        }
    }
    private val json = Json { ignoreUnknownKeys = true }
    private val sessionMutex = Mutex()
    @Volatile private var session: BackendSession? = store.load()
    val userId: String? get() = session?.userId

    suspend fun ensureAnonymousSession() = sessionMutex.withLock {
        if (session != null) return@withLock
        val response = execute("/v1/auth/sign-in/anonymous", "POST", "{}".toByteArray(), authenticated = false)
        captureSession(response)
    }

    suspend fun signIn(email: String, password: String): String = sessionMutex.withLock {
        val body = buildJsonObject { put("email", email); put("password", password) }
        captureSession(execute("/v1/auth/sign-in/email", "POST", body.toString().toByteArray(), authenticated = false))
        checkNotNull(session).userId
    }

    suspend fun exchangeLegacySession(accessToken: String, anonymous: Boolean): String = sessionMutex.withLock {
        val body = buildJsonObject { put("access_token", accessToken) }
        val response = execute("/v1/session/exchange", "POST", body.toString().toByteArray(), authenticated = false)
        val value = json.parseToJsonElement(response.bytes.decodeToString()).jsonObject
        val id = UUID.fromString(value.getValue("user_id").jsonPrimitive.content).toString()
        persist(BackendSession(value.getValue("token").jsonPrimitive.content, id, anonymous))
        id
    }

    suspend fun signOut(localOnly: Boolean = false) = sessionMutex.withLock {
        if (!localOnly && session != null) execute("/v1/auth/sign-out", "POST", "{}".toByteArray())
        store.clear()
        session = null
    }

    suspend fun request(path: String, method: String = "GET", body: JsonElement? = null, query: Map<String, String> = emptyMap()): JsonElement {
        val response = execute(path, method, body?.toString()?.toByteArray(), query = query)
        return json.parseToJsonElement(response.bytes.decodeToString())
    }

    suspend fun upload(bytes: ByteArray, contentType: String, query: Map<String, String>): JsonElement {
        val response = execute("/v1/files/upload", "POST", bytes, contentType = contentType, query = query)
        return json.parseToJsonElement(response.bytes.decodeToString())
    }

    suspend fun download(bucket: String, path: String): ByteArray =
        execute("/v1/files/read", "GET", query = mapOf("bucket" to bucket, "path" to path)).bytes

    private suspend fun execute(
        path: String,
        method: String,
        bytes: ByteArray? = null,
        contentType: String = "application/json",
        query: Map<String, String> = emptyMap(),
        authenticated: Boolean = true,
    ): BackendResponse = withContext(Dispatchers.IO) {
        require(path.startsWith("/v1/")) { "Invalid versioned API path" }
        val url = originUrl.newBuilder().encodedPath(path).apply {
            query.forEach { (key, value) -> addQueryParameter(key, value) }
        }.build()
        val request = Request.Builder().url(url).header("Accept", "application/json")
            .header("Cache-Control", "no-store")
        if (authenticated) {
            val active = session ?: throw BackendException(401, "unauthenticated", "请重新登录。", null)
            request.header("Authorization", "Bearer ${active.token}")
        }
        val body = bytes?.toRequestBody(contentType.toMediaType())
            ?: if (method in listOf("POST", "PUT", "PATCH")) ByteArray(0).toRequestBody(contentType.toMediaType()) else null
        http.newCall(request.method(method, body).build()).execute().use { response ->
            val data = response.body?.bytes() ?: ByteArray(0)
            if (!response.isSuccessful) {
                val error = runCatching { json.parseToJsonElement(data.decodeToString()).jsonObject }.getOrNull()
                val envelope = error?.get("errorEnvelope")?.jsonObject
                throw BackendException(
                    response.code,
                    envelope?.get("code")?.jsonPrimitive?.content ?: error?.get("code")?.jsonPrimitive?.content ?: "http_${response.code}",
                    envelope?.get("message")?.jsonPrimitive?.content ?: error?.get("error")?.jsonPrimitive?.content ?: error?.get("message")?.jsonPrimitive?.content ?: "后台请求失败，请稍后重试。",
                    response.header("X-Request-ID"),
                )
            }
            BackendResponse(data, response.header("set-auth-token"))
        }
    }

    private fun captureSession(response: BackendResponse) {
        val token = response.signedToken?.takeIf(String::isNotBlank)
            ?: throw BackendException(0, "missing_signed_session", "登录状态未建立，请重试。", null)
        val user = json.parseToJsonElement(response.bytes.decodeToString()).jsonObject.getValue("user").jsonObject
        val id = UUID.fromString(user.getValue("id").jsonPrimitive.content).toString()
        persist(BackendSession(token, id, user["isAnonymous"]?.jsonPrimitive?.content == "true"))
    }

    private fun persist(value: BackendSession) {
        store.save(value)
        session = value
    }

    private data class BackendResponse(val bytes: ByteArray, val signedToken: String?)
}
